Efficient backslash solve for Kronecker systems

SebastianAment opened the pull request:
I implemented a backslash method for Kronecker systems. 
You can test the speed by trying 
```
n = 32
m = 32
A = randn(n,n);
B = randn(m,m);
K = kronecker(A,B);
c = randn(n*m);
M = Matrix(K);
@time K\c;
@time M\c;
```
Compared to the naive method, I get a speed-up of 2-3 orders of magnitude on my computer for this example.

willtebbutt commented on src/base.jl:
```diff
@@ -307,3 +307,13 @@ function Base.:*(K::AbstractKroneckerProduct, a::Number)
+# SOLVING
+using LinearAlgebra: checksquare
+function LinearAlgebra.:\(K::AbstractKroneckerProduct{T}, c::AbstractVector{T}) where {T}
+    !issquare(K.A) && throw(DimensionMismatch("matrix A is not square: dimensions are " * size(K.A)))
```
Perhaps this should fallback to a slow method rather than bailing out entirely if a non-square matrix is encountered?

SebastianAment replied:
You're right, I wasn't thinking of least-squares. I'll make the change.

SebastianAment replied:
I just updated it. Let me know if anything else comes up.

SebastianAment commented:
I realized we can also use the efficient backslash for least squares problems (with non-square A,B). I deleted the square-checking, and modified the non-square test, so that `Kx = c` does not have a solution. The test now shows that the normal equations are solved, providing the least-squares solution. 

A related issue: I saw that many other methods in this project throw errors if the component matrices are not square. For example, it is mathematically correct for `det` to return zero on a square Kronecker matrix with non-square component matrices. Instead, it throws an error.

MichielStock commented:
This is awesome work!
